homeassistant/components/idasen_desk/switch.py

```diff
@@ -0,0 +1,117 @@
+    )
+
+
+class IdasenDeskSwitch(CoordinatorEntity, SwitchEntity, RestoreEntity):
+    """Idasen Desk connection switch."""
+
+    _attr_device_class = SwitchDeviceClass.SWITCH
+
+    def __init__(
+        self,
+        desk: Desk,
+        address: str,
+        device_info: DeviceInfo,
+        coordinator: DataUpdateCoordinator,
+    ) -> None:
+        """Initialize the connection switch."""
+        super().__init__(coordinator)
+        self._desk = desk
+        self._address = address
+        self._attr_name = f"{device_info.get(ATTR_NAME)} Connection"
+        self._attr_unique_id = address
+        self._attr_device_info = device_info
+        self._attr_is_on = False
+
+    async def _connect(self) -> None:
+        """Connect and handle connection failures."""
+        _LOGGER.debug("Trying to connect %s", self._address)
+        ble_device = bluetooth.async_ble_device_from_address(
+            self.hass, self._address, connectable=True
+        )
+        if ble_device is None:
+            await self.async_turn_off()
+            return
+        await self._desk.connect(ble_device)
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+        last_state = await self.async_get_last_state()
+        if last_state and last_state.state == STATE_OFF:
+            await self.async_turn_off()
+        else:
+            # If last state is None it should also turn on, since it should be
+            # on by default unless the user has explicitly turned it off.
+            await self.async_turn_on()
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use in the frontend."""
+        if not self.is_on:
+            return "mdi:bluetooth-off"
+        if self._desk.is_connected:
+            return "mdi:bluetooth-connect"
+        return "mdi:bluetooth"
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn connection on."""
+        _LOGGER.debug("Turn Idasen Desk connection on %s", self._address)
+        self._attr_is_on = True
+        self.async_write_ha_state()
+        await self._connect()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn connection off."""
+        _LOGGER.debug("Turn Idasen Desk connection off %s", self._address)
+        self._attr_is_on = False
+        await self._desk.disconnect()
+        self.async_write_ha_state()
+
+    @callback
+    def _handle_coordinator_update(self, *args: Any) -> None:
+        """Handle data update."""
+        if self.is_on:
+            if not self._desk.is_connected:
+                self.hass.async_create_task(self._connect())
+        elif self._desk.is_connected:
+            _LOGGER.warning("Desk is connected but should not be. Disconnecting now")
+            self.hass.async_create_task(self._desk.disconnect())
+
+        self.async_write_ha_state()
```

Comment:
It would be better to flip a flag and let the coordinator handle this as its a bit of a seperation of concerns issue to have it inside the entity

Reply:
Makes sense!